src/aiida_quantumespresso/parsers/base.py

```diff
@@ -50,14 +113,96 @@ def emit_logs(self, logging_dictionaries, ignore=None):
                     except AttributeError:
                         pass
 
-    def exit(self, exit_code):
+    def check_base_errors(self, logs: AttributeDict) -> Optional[ExitCode]:
+        """Check the ``logs`` for the following "basic" parsing error and return a (formatted) version:
+
+        * ``ERROR_OUTPUT_STDOUT_MISSING``
+        * ``ERROR_OUTPUT_STDOUT_READ``
+        * ``ERROR_OUTPUT_STDOUT_PARSE``
+
+        These errors mean that there is no ``stdout`` to parse.
+
+        The ``ERROR_OUTPUT_STDOUT_INCOMPLETE`` error is not checked here because in this case there might still be
+        useful information in the ``stdout``.
+        """
+
+        for exit_code in [
+            'ERROR_OUTPUT_STDOUT_MISSING',
+        ]:
+            if exit_code in logs.error:
+                return self.exit_codes.get(exit_code)
+
+        # These exit codes have additional information that needs to be formatted in the message.
+        for exit_code in [
+            'ERROR_OUTPUT_STDOUT_READ',
+            'ERROR_OUTPUT_STDOUT_PARSE'
+        ]:
+            if exit_code in logs.error:
+                exception = logs.error[logs.index(exit_code) + 1]
+                return self.exit_codes.get(exit_code).format(exception=exception)
+
+    def exit(self, exit_code: ExitCode, logs: AttributeDict = None) -> ExitCode:
         """Log the exit message of the give exit code with level `ERROR` and return the exit code.
 
         This is a utility function if one wants to return from the parse method and automically add the exit message
-        associated to the exit code as a log message to the node: e.g. `return self.exit(self.exit_codes.LABEL))`
+        associated to the exit code as a log message to the node: e.g. ``return self._exit(self.exit_codes.LABEL))``
 
         :param exit_code: an `ExitCode`
         :return: the exit code
         """
+        if logs:
+            self.emit_logs(logs)
```

Comment:
I was wondering what to do here. I liked the idea of always passing the `logs` to the `exit` method, so the logs are always emitted before the parsing stops. But this lead to a few questions:

1. What does the `emit_logs` method _really_ do? It seems to only use `getattr`, how does it really "emit" anything?
2. What's the purpose of this `logger` exactly? 
3. this reminds me, perhaps we should end each `parse` method with an explicit `return` of a zero exit code via the `exit` method? Explicit is better than implicit, and then it passes through this method which emits the logs (now).

Reply:
> What does the `emit_logs` method _really_ do? It seems to only use `getattr`, how does it really "emit" anything?

Believe it or not, it emits log messages 😄 It is easy to miss but the line
```
getattr(self.logger, level)(stripped)
```
will retrieve a method of the `Logger` instance stored at `self.logger`. Since `level` is a log-level in lower-case, e.g., `warning` or `error`, `getattr(self.logger, level)` will return the bound _method_ `warning` or `error` in that case. We then _call_ it passing the message `...(stripped)`. So essentially it is doing something like `self.logger.warning(stripped)`.

> What's the purpose of this logger exactly?

It is configured with the associated `CalcJobNode` whose outputs are being parsed and so any log messages are stored as `Log` entries linked to the node. They will show up in the output of `verdi process report` for example. This is useful information for the user.

> this reminds me, perhaps we should end each parse method with an explicit return of a zero exit code via the exit method? Explicit is better than implicit, and then it passes through this method which emits the logs (now).

Yes, but we should then also remember to first check whether an exit status was already set by the scheduler plugin. For example if the scheduler parser detected the job was killed due to an OOW error, the associated exit status (I believe 120) is already set on the node. So before we return `ExitCode(0)`, we should check if `self.node.exit_status is not None` in which case we probably want to return `None` to keep that status, because returning `ExitCode(0)` will override it.

Reply:
Great, thanks for the clarification, @sphuber! 🙏 

> So essentially it is doing something like self.logger.warning(stripped).

Right, of course, it's quite clear after this explanation. I guess for this kind of use case, a `CalcJobNode.get_logger(level)` method would be more readable than a `logger` property. ^^

Now that I understand better, I'd be inclined to merge the `emit_logs` and `exit` methods into one method with signature:

```python
def exit(self, logs: AttributeDict = None, exit_code: ExitCode = None) -> ExitCode:
```

which:

1. Emits any logs left in the `logs` input, if provided.
2. If `exit_code` has been provided, emit its `message` via `self.logger`.
3. If not, check if there is already an `exit_status` set on the node and log this `message` (or will this have been done already?)
4. Return 
   * `exit_code` if it was provided.
   * `None` if the node already has an `exit_status`.
   * `ExitCode(0)` else

Reply:
> Right, of course, it's quite clear after this explanation. I guess for this kind of use case, a `CalcJobNode.get_logger(level)` method would be more readable than a `logger` property. ^^

Don't think that would be a correct name, because you cannot return a logger for a particular level, you log a message at a particular level through a logger. That being said, I think you _can_ easily make the current code more readable. I think the `Logger` class has a `log` method where you can pass the desired level as an argument, so that line would become `self.logger.log(level, stripped)`. Pretty sure that should work.

> Now that I understand better, I'd be inclined to merge the emit_logs and exit methods into one method with signature:

I would keep `emit_logs` as a separate function, because it can be useful to call it directly elsewhere still. Fine to move the call in all parsers to the `exit` method.

> If not, check if there is already an exit_status set on the node and log this message (or will this have been done already?)

This will already have been done by the engine before calling the `Parser`. So nothing to do here but to "keep" it if the parser itself cannot provide a more specific exit code.

Reply:
I've adapted the method as discussed, see 6529ea6d22cf532ee07496a97c66838d987b35f7.

> I would keep emit_logs as a separate function, because it can be useful to call it directly elsewhere still. Fine to move the call in all parsers to the exit method.

If we _always_ end the `parse` method with the `exit` method for clarity, and always emit _all_ the logs there, what would be the use case of calling the `emit_logs` method separate?